src/ir/auth_logic/souffle_emitter.h

```diff
@@ -102,14 +120,48 @@ class SouffleEmitter {
     return absl::StrCat(".decl ", predicate.name(), "(", arguments, ")");
   }
 
-  std::string EmitDeclarations() {
-    std::vector elements_vec(std::make_move_iterator(declarations_.begin()),
-                             std::make_move_iterator(declarations_.end()));
-    std::sort(elements_vec.begin(), elements_vec.end());
-    return absl::StrJoin(elements_vec, "\n",
-                         [this](std::string* out, auto decl) {
-                           return absl::StrAppend(out, EmitDeclaration(decl));
-                         });
+  std::string EmitArguments(std::vector<datalog::Argument> arguments) {
+    return absl::StrJoin(
+        arguments, ", ", [](std::string* out, datalog::Argument argument) {
+          return absl::StrAppend(out, argument.argument_name(), " : ",
+                                 argument.argument_type().name());
+        });
+  }
+
+  std::string EmitRelationDeclarations(
+      const datalog::Program& program,
+      const std::vector<std::string>& skip_declarations) {
+    std::vector<std::string> declaration_strings;
+    for (const auto& declaration : program.relation_declarations()) {
+      if (absl::c_find(skip_declarations, declaration.relation_name()) !=
+          skip_declarations.end())
+        continue;
+      declaration_strings.push_back(
+          absl::StrCat(".decl ", declaration.relation_name(), "(",
+                       EmitArguments(declaration.arguments()), ")"));
+    }
+    std::sort(declaration_strings.begin(), declaration_strings.end());
+    return absl::StrJoin(declaration_strings, "\n");
+  }
+
+  std::string EmitTypeDeclarations(
+      const datalog::Program& program,
+      const std::vector<std::string>& skip_declarations) {
+    std::vector<std::string> type_names;
+    for (const auto& declaration : program.relation_declarations()) {
+      for (const auto& argument : declaration.arguments()) {
+        if (argument.argument_type().kind() !=
+            datalog::ArgumentType::Kind::kCustom)
+          continue;
+        if (absl::c_find(skip_declarations, argument.argument_name()) !=
+            skip_declarations.end())
+          continue;
```

Comment:
Again, unnecessarily linear. Change `skip_declarations` to be an `absl::flat_hash_set` to cut down the asymptotic time.

Reply:
skip_declarations is already a flat_hash_set, change to `skip_declarations.find` ?